### PR TITLE
Added documentation about query composition through fragments

### PR DESCRIPTION
**Pull request by Poincare**
Added documentation about a recently-added feature in Apollo Client: doing query composition by referencing fragments.

### CODE REVIEW

File: source/apollo-client/core.md [hunk 1]
```diff
@@ -129,6 +129,7 @@ Run a GraphQL query and return a promise that resolves to a `GraphQLResult`.
 - `query: string` A GraphQL query string to fetch.
 - `variables: Object` The variables to pass along with the query.
 - `forceFetch: boolean` (Optional, default is `false`) If true, send the query to the server directly without any pre-processing. If false, check if we have some of the data for the query on the client already, and send a minimized query to the server to refetch only the objects we don't have already.
+- `fragments: FragmentDefinition[]` (Optional, default is `[]`) An array of fragment definitions that the `query` can reference. Generally created through `createFragment`, as described in the [fragments section](core.html#fragments).
```
Comment (stubailo) [inline source/apollo-client/core.md]: I think this shouldn't talk about fragment definitions - it can just say "an array of fragments created by `createFragment`".

File: source/apollo-client/core.md [hunk 2]
```diff
@@ -317,3 +320,70 @@ Right now, this is a bit verbose because you have to list the names of the varia
...
+
+Apollo Client supports composition of queries and mutations using named fragments. For example, if you have several UI components and each defines a fragment that it must fetch in order to render itself, a root UI component can compose a single query that references each of those fragments. This single query is then fetched from the server and will return the data requested by each of the fragments. This alows you to get the data needed to render all of those UI components in a single request to your GraphQL server.
+
+For query composition through fragments to work, Apollo Client requires unique fragment names across your application and will warn you if this is not the case. We can use the `createFragment` method to register a fragment for use in queries or mutations.
+
+<h3 id="createFragment" title="createFragment">createFragment</h3>
+
+To create a fragment, you must use the `createFragment` method which is exposed as a named export of `apollo-client`. It keeps track of the fragment names being used across your application and will print a warning if there is a naming collision across a pair of named fragments. It takes the following arguments:
```
Comment (stubailo) [inline source/apollo-client/core.md]: This should have an `import { createFragment } from 'apollo-client'` code snippet somewhere.

File: source/apollo-client/core.md [hunk 3]
```diff
@@ -317,3 +320,70 @@ Right now, this is a bit verbose because you have to list the names of the varia
...
+
+- `doc: Document`: A GraphQL document represented as a GraphQL AST. This should contain the fragment definitions you would like use within your queries.
```
Comment (stubailo) [inline source/apollo-client/core.md]: I bet most people using this aren't going to be aware of GraphQL ASTs. we should frame it as something returned by `gql`. Perhaps we should lead with a code snippet.

File: source/apollo-client/core.md [hunk 4]
```diff
@@ -317,3 +320,70 @@ Right now, this is a bit verbose because you have to list the names of the varia
...
+  query {
+    publisher {
+      name
+    }
+  }`,
+})
+
+```
+
+Instead of firing those two queries separately, we can define the data that they request as fragments and register them with Apollo Client. Then, we can compose a single query that uses these fragments and fire that instead. We can do that with `createFragment`:
```
Comment (stubailo) [inline source/apollo-client/core.md]: I don't think this is a good example since this is something Apollo Client would do for you anyway. A good use of fragments is specifically when you have nested data, or you need to share a fragment between a query and mutation. Or, if you have a UI component that declares its own fields.

**Comment by Poincare** [comment]
Fixed @stubailo

File: source/apollo-client/core.md [hunk 5]
```diff
@@ -317,3 +320,77 @@ Right now, this is a bit verbose because you have to list the names of the varia
...
 
 Where it is not supported, you can use one of several popular polyfills, including [whatwg-fetch](https://github.com/github/fetch), [node-fetch](https://github.com/bitinn/node-fetch) or [isomorphic-fetch](https://github.com/matthew-andrews/isomorphic-fetch).
+
+<h2 id="fragments" title="Fragments">Fragments</h2>
+
+Apollo Client supports composition of queries and mutations using named fragments. For example, if you have several UI components and each defines a fragment that it must fetch in order to render itself, a root UI component can compose a single query that references each of those fragments. This single query is then fetched from the server and will return the data requested by each of the fragments. This alows you to get the data needed to render all of those UI components in a single request to your GraphQL server.
+
+For query composition through fragments to work, Apollo Client requires unique fragment names across your application and will warn you if this is not the case. We can use the `createFragment` method to register a fragment for use in queries or mutations.
```
Comment (stubailo) [inline source/apollo-client/core.md]: I'd rephrase this to make it sound less like something that Apollo Client imposes on you because of a technical limitation, but more of a feature so that it's easier to debug queries in your server logs.

File: source/apollo-client/core.md [hunk 6]
```diff
@@ -317,3 +320,77 @@ Right now, this is a bit verbose because you have to list the names of the varia
...
+    author {
+      ...authorNames
+      ...authorCities
+    }
+  }
+  `,
+}, fragmentDefs);
+```
+
+By doing this, we're no longer firing two queries. Instead, we're only firing one query which will get us all the data that we need. So, query composition through named fragments benefits you by reducing the number of roundtrips you make to the server. Query composition through fragments is often useful when we you want to load nested data, refetch the same data in a mutation and a query or if a UI component defines its own fields.
```
Comment (stubailo) [inline source/apollo-client/core.md]: I still don't think this is a good example. A better one would be something like:

```
{
  author(id: 5) {
    books {
      id
      name
    }
  }
}
```

And then,

```
{
  book(id: 19) {
    coverImage
    numPages
  }
}
```

So that's a case where it would be much better, instead of doing those queries in series (waiting for the first to come back), to do a fragment:

```
{
  author(id: 5) {
    books {
      id
      name
      ...bookInfo
    }
  }
}
```

**Comment by stubailo** [comment]
I wrote some stuff here: http://docs.apollostack.com/apollo-client/fragments.html
